### PR TITLE
DevTools: Check if string conversion is possible via TypeConverterter and fallback to ToString()

### PR DESCRIPTION
**Pull request by pr8x**
This PR refactors DevTools to prefer user-provided `ToString()` method over generic `TypeDescriptor`. This makes inspecting `ColumnDefinition` (and some other types) a bit more helpful.

image
image

### CODE REVIEW

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs [hunk 1]
```diff
@@ -25,19 +26,37 @@ protected static string ConvertToString(object value)
...
+            //Check if there's an user provided ToString(), prefer that over the TypeDescriptor conversion
+            if (value.GetType().GetMethod(nameof(ToString), System.Type.EmptyTypes)
+                .DeclaringType != typeof(object))
+            {
+                return value.ToString();
+            }
+
+            try
+            {
+                var converter = TypeDescriptor.GetConverter(value);
```
Comment (pr8x) [inline src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:38]: Note that `GetConverter()` never returns null, but rather throws or returns a default converter.

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs [hunk 2]
```diff
@@ -25,19 +26,37 @@ protected static string ConvertToString(object value)
                 return "(null)";
             }
 
-            var converter = TypeDescriptor.GetConverter(value);
-            return converter?.ConvertToString(value) ?? value.ToString();
+            //Check if there's an user provided ToString(), prefer that over the TypeDescriptor conversion
+            if (value.GetType().GetMethod(nameof(ToString), System.Type.EmptyTypes)
```
Comment (kekekeks) [inline src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:30]: This kinda conflicts with zero reflection long-time goal

Comment (MarchingCube) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:30]: Should this apply to DevTools though? It kinda relies on reflection to do type introspection, we even have console there.

We should definitely avoid using reflection in core framework but tooling might need to be exempt from this. Opinions?

Comment (kekekeks) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:29]: I'm still planning to make devtools useful when app is running on iOS device. So there will be a no-JIT requirement at some point.

Comment (MarchingCube) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:29]: I wonder if we could have such features pluggable, so we can have fully featured DevTools on desktop and a stripped version on platforms that don't support JIT/Reflection/Etc.

Comment (pr8x) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:29]: How do you want to run DevTools on mobile devices?

**Comment by grokys** [comment]
Not sure about this; feels like a hack - how do we know that if the developer has overridden `ToString` that they did it to make the string more useful to devtools? What if the developer decided to improve the type converter to make it more useful than `ToString`?

To me the obvious way to fix this feels like it would be to fix the `TypeConverter`s string output?

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs (hunk 2 above)
Comment (kekekeks) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs:29]: The same way our previewer currently works on Rider - by listening on a TCP socket and providing a small web page that loads frames and sends input via websocket. The infrastructure is already there.

**Comment by pr8x** [comment]
@grokys IDK I figured that most people actually use `ToString()` as it's much easier to write and produces less boilerplate code. But I guess I can also fix the TypeConverter for `ColumnDefinition` / `RowDefinition`.

**Comment by grokys** [comment]
@pr8x yeah I understand, not sure myself what the best approach is here. Is it possible to detect a default type converter? If so maybe we could detect a default type converter and if so use `ToString()`? Or does the default type converter just call `ToString`?

**Comment by pr8x** [comment]
Okay updated it to only filter out `CollectionConverter` (which is really useless - see https://github.com/dotnet/runtime/blob/2ba65159b970de1af69e350bf8ac67ab4c21e38b/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/CollectionConverter.cs#L20) and fallback to `ToString()`. Also reflection parsing should now work correctly when the given `TypeConverter` does not support conversion from string.

**Comment by Gillibald** [comment]
Can we not just support https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debuggerdisplayattribute?view=netcore-3.1 ?

**Comment by pr8x** [comment]
@Gillibald I guess we can check if it's available and prefer it over everything else.

File: src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs (hunk 2 above)
Comment (pr8x) [reply src/Avalonia.Diagnostics/Diagnostics/ViewModels/PropertyViewModel.cs]: The reflection is gone :)

**Comment by Gillibald** [comment]
@pr8x https://github.com/dotnet/roslyn/blob/master/src/Scripting/CSharp/Hosting/ObjectFormatter/CSharpObjectFormatter.cs but that adds a dependency

**Comment by pr8x** [comment]
@Gillibald Do you think it makes sense to include that dependency for this? If yes, I can integrate it.

**Comment by Gillibald** [comment]
We don't use DebuggerDisplayAttribute so it's fine to leave that part out. We can add it if needed later in time.

**Review by grokys — APPROVED** [review]
Thanks, looks good now!

**Review by MarchingCube — APPROVED** [review]
LGTM.

**Pull request merged**